circle-mlir/circle-mlir/lib/dialect/src/ops/UnsqueezeOnnxOp.h

```diff
@@ -0,0 +1,171 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __CIRCLE_MLIR_DIALECT_OPS_UNSQUEEZE_ONNX_OP_H__
+#define __CIRCLE_MLIR_DIALECT_OPS_UNSQUEEZE_ONNX_OP_H__
+
+#include "circle-mlir/dialect/CircleDialect.h"
+#include "circle-mlir/dialect/NameUtils.h"
+
+#include <mlir/IR/Matchers.h>
+#include <mlir/IR/PatternMatch.h>
+
+namespace mlir
+{
+namespace Circle
+{
+
+//===----------------------------------------------------------------------===//
+// UnsqueezeOnnxOp: temporary Op for conversion
+//===----------------------------------------------------------------------===//
+
+namespace
+{
+
+bool valididateAxesValues(mlir::RankedTensorType &intype, mlir::RankedTensorType &outtype,
+                          std::vector<int32_t> &axesValues)
+{
+  if (axesValues.size() == 0)
+  {
+    LLVM_DEBUG({ llvm::dbgs() << "UnsqueezeOnnxOp2ResizeOp axesValues none\n"; });
+    return false;
+  }
+
+  // NOTE order of values in axes does not matter and can come in any order.
+  auto inshape = intype.getShape();
+  int32_t inshapeSize = static_cast<int32_t>(inshape.size());
+  const int32_t outSize = inshapeSize + static_cast<int32_t>(axesValues.size());
+  for (int32_t i = 0; i < axesValues.size(); ++i)
+  {
+    int32_t value = axesValues[i];
+    axesValues[i] = value < 0 ? value + outSize : value;
+  }
+  sort(axesValues.begin(), axesValues.end());
+
+  // NOTE input axes should not contain any duplicate entries
+  if (std::adjacent_find(axesValues.begin(), axesValues.end()) != axesValues.end())
+  {
+    LLVM_DEBUG({ llvm::dbgs() << "UnsqueezeOnnxOp2ResizeOp axesValues duplicate\n"; });
+    return false;
+  }
+
+  return true;
+}
+
+mlir::Value getUnsqueezeShape(mlir::PatternRewriter &rewriter, mlir::RankedTensorType &intype,
+                              mlir::RankedTensorType &outtype, std::string name,
+                              std::vector<int32_t> &axesValues)
+{
+  // 1) We can ignore axes and use outtype shape
+  // 2) We can produce shape with intype shape + axes and compare with outtype shape
+  // --> go with 2
+  auto inshape = intype.getShape();
+  int32_t inshapeSize = static_cast<int32_t>(inshape.size());
+
+  std::vector<int32_t> values;
+  for (int32_t i = 0; i < inshapeSize; ++i)
+    values.push_back(static_cast<int32_t>(inshape[i]));
+  for (size_t i = 0; i < axesValues.size(); ++i)
+  {
+    auto idx = axesValues.at(i);
+    values.insert(values.begin() + idx, 1);
+  }
+
+  // expand to match output shape
+  while (values.size() < static_cast<size_t>(inshapeSize) + axesValues.size())
+    values.push_back(1);
+
+  // verify that shape values should be same as output shape
+  auto outshape = outtype.getShape();
+  assert(outshape.size() == values.size());
+  for (size_t i = 0; i < outshape.size(); ++i)
+  {
+    // refer https://github.sec.samsung.net/one-project/circle-mlir/issues/871#issuecomment-2159296
+    if (values[i] == 0 && outshape[i] < 0)
```

Comment:
Let's remove the inner URL.

```suggestion
  {
    if (values[i] == 0 && outshape[i] < 0)
```